src/peer_manager.rs

```diff
@@ -296,17 +311,184 @@ impl PeerManager {
+            let mut rng = SeededRng::new();
+            let seed = rng.gen_iter().take(10).collect();
+            self.node_candidate = Some((*expected_name, Instant::now(), seed));
+        } else {
+            return Ok((false, (prefix, vec![])));
+        }
+
+        let mut public_ids = vec![];
+        for name in names {
+            if name == *our_public_id.name() {
+                public_ids.push(*our_public_id);
+            } else if let Some(peer) = self.peer_map.get_by_name(&name) {
+                public_ids.push(*peer.pub_id())
+            }
+        }
+        public_ids.sort();
+
+        Ok((true, (prefix, public_ids)))
+    }
+
+    pub fn add_as_peer_candidate(&mut self, name: XorName) {
+        self.peer_candidates.push(name)
+    }
+
+    pub fn verify_candidate(&mut self,
+                            node_name: XorName,
+                            proof: Vec<u8>,
+                            _leading_zero_bytes: u32,
+                            _difficulty: u32)
+                            -> Option<bool> {
+        if let Some((ref name, ref now, ref seed)) = self.node_candidate {
+            if *name == node_name &&
+               now.elapsed() < Duration::from_secs(RESOURCE_PROOF_EVALUATE_TIMEOUT_SECS) {
+                return Some(*seed == proof);
+            }
+        }
+        None
+    }
+
+    pub fn handle_node_approval_vote(&mut self,
+                                     candidate_name: XorName,
+                                     validity: bool)
+                                     -> (bool, Option<(PublicId, PeerId)>) {
+        if let Some((name, _, _)) = self.node_candidate {
+            if name == candidate_name {
+                let result = if validity {
+                    let (pub_id, peer_id) = if let Some(peer) = self.peer_map
+                        .get_by_name(&candidate_name) {
+                        if let Some(peer_id) = peer.peer_id().clone() {
+                            (*peer.pub_id(), *peer_id)
+                        } else {
+                            return (false, None);
+                        }
+                    } else {
+                        return (false, None);
+                    };
+                    Some((pub_id, peer_id))
+                } else {
+                    None
+                };
+                self.node_candidate = None;
+                return (true, result);
+            }
+        }
+        (false, None)
+    }
+
+    pub fn handle_node_approval(&mut self) -> Vec<(PublicId, PeerId)> {
+        let mut result = vec![];
+        for peer_name in &self.peer_candidates.clone() {
+            if let Some(peer) = self.peer_map.get_by_name(peer_name) {
+                if let Some(peer_id) = peer.peer_id().clone() {
+                    result.push((*peer.pub_id(), *peer_id));
+                }
+            }
+        }
+        self.peer_candidates.clear();
+        result
+    }
+
+    pub fn is_node_candidate(&mut self,
```

Comment:
I would assume a method named `is_...` would just check whether some property is true, but this also updates the peer state. Could you at least add documentation of this, if not also find a better name?